contracts/marketpools.js

```diff
@@ -0,0 +1,346 @@
+    || !api.assert(typeof (tradeType) === 'string' && TradeType.indexOf(tradeType) !== -1, 'invalid tradeType')
+    || !await validateTokenPair(tokenPair)) {
+    return;
+  }
+
+  const [baseSymbol, quoteSymbol] = tokenPair.split(':');
+  const pool = await api.db.findOne('pools', { tokenPair });
+  if (!api.assert(pool, 'no existing pool for tokenPair')) return;
+  let liquidityIn;
+  let liquidityOut;
+  let symbolIn;
+  let symbolOut;
+  const tradeDirection = tradeType === 'exactInput' ? tokenSymbol === baseSymbol : tokenSymbol !== baseSymbol;
+  if (tradeDirection) {
+    liquidityIn = pool.baseQuantity;
+    liquidityOut = pool.quoteQuantity;
+    symbolIn = baseSymbol;
+    symbolOut = quoteSymbol;
+  } else {
+    liquidityIn = pool.quoteQuantity;
+    liquidityOut = pool.baseQuantity;
+    symbolIn = quoteSymbol;
+    symbolOut = baseSymbol;
+  }
+
+  const tokenIn = await api.db.findOneInTable('tokens', 'tokens', { symbol: symbolIn });
+  const tokenOut = await api.db.findOneInTable('tokens', 'tokens', { symbol: symbolOut });
+
+  const senderBase = await api.db.findOneInTable('tokens', 'balances', { account: api.sender, symbol: symbolIn });
+  let senderBaseFunded = false;
+  let tokenPairDelta;
+  let tokenQuantity;
+  if (tradeType === 'exactInput') {
+    const tokenAmountAdjusted = api.BigNumber(getAmountOut(tokenAmount, liquidityIn, liquidityOut));
+    if (!tokenAmountAdjusted.isFinite()) return;
+    senderBaseFunded = senderBase && api.BigNumber(senderBase.balance).gte(tokenAmount);
+    tokenPairDelta = tokenSymbol === baseSymbol ? [tokenAmount, api.BigNumber(tokenAmountAdjusted).times(-1)] : [api.BigNumber(tokenAmountAdjusted).times(-1), tokenAmount];
+    tokenQuantity = { in: tokenAmount, out: tokenAmountAdjusted };
+    if (!api.assert(api.BigNumber(tokenQuantity.in).dp() <= tokenIn.precision, 'symbolIn precision mismatch')) return;
+  } else if (tradeType === 'exactOutput') {
+    const tokenAmountAdjusted = api.BigNumber(getAmountIn(tokenAmount, liquidityIn, liquidityOut));
+    if (!tokenAmountAdjusted.isFinite()) return;
+    senderBaseFunded = senderBase && api.BigNumber(senderBase.balance).gte(tokenAmountAdjusted);
+    tokenPairDelta = tokenSymbol === baseSymbol ? [api.BigNumber(tokenAmount).times(-1), tokenAmountAdjusted] : [tokenAmountAdjusted, api.BigNumber(tokenAmount).times(-1)];
+    tokenQuantity = { in: tokenAmountAdjusted, out: tokenAmount };
+    if (!api.assert(api.BigNumber(tokenQuantity.out).dp() <= tokenOut.precision, 'symbolOut precision mismatch')) return;
+  }
+
+  if (!api.assert(senderBaseFunded, 'insufficient input balance')
+    || !validateSwap(pool, tokenPairDelta[0], tokenPairDelta[1], api.BigNumber(maxSlippage).dividedBy(100))) return;
+
+  const res = await api.executeSmartContract('tokens', 'transferToContract', { symbol: symbolIn, quantity: api.BigNumber(tokenQuantity.in).toFixed(tokenIn.precision), to: 'marketpools' });
```

Comment:
The senderBaseFunded check is done on the unrounded amount.  The final amount send by the the transferToContract action may be slightly different due to the rounding by toFixed.  You should modify the senderBaseFunded check to be done on the final amount to be sent (with the toFixed rounding).

Reply:
I don't think this would be an issue now that we've put ROUND_DOWN in most places but I've added it anyway to be safe.